src/structs.py

```diff
@@ -740,17 +740,66 @@ def unset_option(self) -> None:
         assert not self.has_option()
 
 
+@dataclass(frozen=True, repr=False, eq=False)
```

Comment:
I like that these are frozen, that means we can probably do some caching that wasn't possible before...